### ISSUE TITLE
After macos wakeup, the system usb setting  increase one sd card item

### ISSUE THREAD

**Issue by nabaonan** [0xFireWolf/RealtekCardReader#22]
- Sleep 
- Wake up
- See system setting  usb  there will be added a sd item which called CRW

**Comment by nabaonan** [comment]
![image](https://user-images.githubusercontent.com/17247526/140449768-a25731cd-e18a-42df-acfa-3d3758c99358.png)

**Comment by nabaonan** [comment]
![image](https://user-images.githubusercontent.com/17247526/140600424-f338515f-de04-4eda-8b23-08e9595492b0.png)

**Comment by nabaonan** [comment]
[rtsx.log](https://github.com/0xFireWolf/RealtekCardReader/files/7490325/rtsx.log)

**Comment by parnz** [comment]
I also have this problem and it doesn't detect sd card when insert at all.

![Screen Shot 2021-12-09 at 19 25 39](https://user-images.githubusercontent.com/1014336/145396400-f060d435-88ce-477b-bdc6-cfbc34e2c167.png)

**Comment by dblhlx** [comment]
I have this problem too. The card is RTS5129.

**Comment by 0xFireWolf** [comment]
This is a known issue of RTS5129 (probably RTS5139 and RTS5179 as well).

My suspicion is that the USB card reader got disconnected when **some** laptops went to sleep, but the driver still assumes that the device remains connected thus holding a reference to it. When your laptop wakes up, the device gets reconnected, so the USB driver stack publishes a new instance of your device, but the old instance is still there because the driver holds a reference to it. Unfortunately, I do not have RTS5129 on my side, so I cannot test my hypothesis. There might be a quick yet dirty fix that disables the power management routines for RTS5129 if a special boot argument is provided, so the driver tears down when your laptop goes to sleep, and a new driver instance starts when your laptop wakes up.

**Comment by dblhlx** [comment]
Thanks so much for you excellent work! I did not expect the card reader will be useful when setting the laptop as a Hackintosh. To my surprise, I can quickly transfer files using the driver you wrote. Your hypothesis is very likely true. My laptop is a Gateway GWTN156-3BK, not a common brand now. I disabled the dGPU (nvidia RTS 2060) using SSDT-dGPU-off.aml, but it always draws significant power after waking up. I need to use SSDT-PTSWAK.aml to turn the dGPU on when the laptop goes to sleep and turn it off when the computer wakes up.

If it is easy to implement your idea, I am more than happy to test your new code. Again thank you very much for your excellent work!

**Comment by 0xFireWolf** [comment]
@dblhlx Thanks for being interested in testing this driver. 

Please try the attached driver and add the boot argument `-rtsxnopm` to disable the power management routine. Please check if the issue persists. 

Specifically, ...
- Do you observe a kernel panic after the laptop wakes up?
- If you open the IORegistryExplorer before the laptop goes to sleep, can you observe a new USB card reader object that shows up in IORegistry after the laptop wakes up?
- Is your USB card reader still working after the laptop wakes up?

Please upload the kernel log dumped after the laptop wakes up as well.

[RealtekCardReader_0.9.7_0a0d816_DEBUG.zip](https://github.com/0xFireWolf/RealtekCardReader/files/7781973/RealtekCardReader_0.9.7_0a0d816_DEBUG.zip)

**Comment by dblhlx** [comment]
I tested the attached kext. Here are the results:
1. No kernel panic was observed after wake-up.
2. A new card reader object showed up in IORegistry after each wake-up (attached).
3. USB card reader still works after wake-up.

Thanks for your excellent work. Please let me know if more testing needed.

![Screen Shot 2021-12-28 at 8 39 57 PM](https://user-images.githubusercontent.com/50333630/147619713-c03dfe48-cf70-4d91-9a10-ba9a13f9f023.png)
[rtsx.log](https://github.com/0xFireWolf/RealtekCardReader/files/7786985/rtsx.log)

**Comment by 0xFireWolf** [comment]
@dblhlx Thanks for the feedback. 

What are the other three `USB2.0-CRW` objects in your screenshot? Did you wake up your laptop only once or for 4 times?

Do you still have the issue that a new `USB2.0-CRW` entry shows up under the USB Device Tree in System Information?
If so, could you please do the following steps and upload the IOReg dump?
- Reboot your computer. Do not insert any SD cards.
- Open the IORegistryExplorer and keep the window open.
- Put your laptop into sleep and wait for one or two minutes.
- Wake up your laptop.
- Save the IOReg dump and upload it.

Thank you.

**Comment by dblhlx** [comment]
In my last post, I waked up my laptop for four times. I opened IORegistryExplorer after the third wake-up.

I followed your instruction this time. I rebooted my laptop and openedIORegistryExplorer, then I put the laptop into sleep. The laptop waked up by itself after a couple of minutes, and I save the IOReg (attached) after that. I don't understand what IOReg dump is. I just did File -> Save as. Please let me know if it is not what you need.

[IORegistry.ioreg.zip](https://github.com/0xFireWolf/RealtekCardReader/files/7787170/IORegistry.ioreg.zip)

**Comment by 0xFireWolf** [comment]
@dblhlx Yes, that's the IOReg dump I need. 

With the boot argument `-rtsxnopm`, the card reader was disconnected and the driver was also terminated when your laptop went to sleep, because all entries under `HS02` were red in the IORegistry. So, it seems that your problem is resolved. Can you confirm that there aren't multiple `USB2.0-CRW` entries under `USB Device Tree` in System Information? Thank you.

**Comment by dblhlx** [comment]
I still see multiple entries of USB2.0-CRW with the boot argument -rtsxnopm. I can also see multiple USB2.0-CRW@14200000/RealtekUSBCardReaderController entries but a single RealtekUSBSDXCSlot entry in the IORegistryExplorer screenshot of my previous post. Nonetheless, I consider this is a cosmetic issue because the card reader is functional.

Thank you very much!
![Screen Shot 2021-12-29 at 7 07 46 AM](https://user-images.githubusercontent.com/50333630/147662628-6889dadc-98a1-4f57-839e-0aca7235179f.png)

**Comment by mac-os-x** [comment]
Hello, I was able to identify the sd card with the same kext, but when I say restart after sleep, the system shuts down too late. I have tested it many times and such a problem occurs in both Big Sur and Monterey. @0xFireWolf

**Comment by mac-os-x** [comment]
![Ekran Resmi 2022-01-24 21 37 27](https://user-images.githubusercontent.com/78235521/150843514-2959aade-1321-4d0b-9171-fc8026456b05.png)

**Comment by prud281** [comment]
-rtsxnopm argument doesn't hel
![Снимок экрана 2022-03-28 в 06 29 22](https://user-images.githubusercontent.com/85940393/160322298-fc2a29a2-0a62-49d0-b20b-9906fc6e516e.png)
p

**Comment by maverx72** [comment]
> -rtsxnopm argument doesn't hel
> ![Снимок экрана 2022-03-28 в 06 29 22](https://user-images.githubusercontent.com/85940393/160322298-fc2a29a2-0a62-49d0-b20b-9906fc6e516e.png)
> p

Same for me

**Comment by maverx72** [comment]
With the boot argument -rtsxnopm there is no change

**Comment by maverx72** [comment]
bluetooth still can't work around 15 minutes, Card reader work, Card reader USB2.0-CRW device still increase and shutdown still slowed after sleep awake operation.

**Comment by toxa70x4** [comment]
Very relevant, waiting for a fix and a new version.
Many thanks to the author.

**Comment by franksfc** [comment]
Is there any chance to fix it? I can offer any information concerned.

**Comment by lslqtz** [comment]
Same issue, I thought it was caused by the system version,so I tried these versions but still had problems. (macOS 11.6.7/macOS 12.0 beta 5/macOS 12.0 beta 6/macOS 12.1/macOS 12.4)

**Comment by bismillah-100** [comment]
I've tried to fix it and the issue has been fixed for me.

[RealtekCardReader_0.9.7__RELEASE-FIX.zip](https://github.com/user-attachments/files/23113730/RealtekCardReader_0.9.7__RELEASE-FIX.zip)

**Comment by bismillah-100** [comment]
Big thanks @0xFireWolf  for this stable kext. The only problem is that some objects cannot be released properly when system prepare to sleep. I tried to fix it and now the object can be released as well as system will go to sleep.

<img width="852" height="785" alt="Image" src="https://github.com/user-attachments/assets/7660580c-3a9b-4056-97d5-0744ab6e7b4f" />

I have modified and tes this kext but not for all cases.

> for most cases, its should be fine like the original one.
> I’ll make a pull request if someone is still here.

[RealtekCardReader_0.9.7_006a845_RELEASE.zip](https://github.com/user-attachments/files/23120761/RealtekCardReader_0.9.7_006a845_RELEASE.zip)

**Comment by 0xFireWolf** [comment]
> > for most cases, its should be fine like the original one.
> > I’ll make a pull request if someone is still here.

@bismillah-100 Thanks for taking the time to investigate the issue. However, I don't have any USB card readers, so I could not test it myself. 

Since you suggested that some objects were not released properly (which is a memory leak), It would be great if you make a pull request anyway, and I will merge it and passively wait for other folks to report problems.
